### ISSUE TITLE
Again about collecting train/test accuracy and loss 

### ISSUE THREAD

**Issue by undertherain** [apache/incubator-mxnet#2873]
I’m tackling the problem of collecting accuracy and loss data on train and validation sets for every epoch. To plot learning curves etc…

There seem to be several related issues here, but mostly addressing minute details.
There’s also a solution suggesting parsing console output(!) to get some useful data, which to me does not seem to be very elegant. 

From what I see it seems possible to collect training accuracy and loss (or whatever metrics are specified) from individual training batch callbacks.
For example each batch_end_callback adds values to some static/global variable and at epoch_callback it is divided by the total number of training batches. 

There’s variable "nbatch" visible among the epoch_callback arguments, do I understand correctly that it is the number of training batches? 

The problem is then with evaluation. First of all, the number of evaluation batches is not explicitly reported. Of course, knowing data size and batch size it is possible to calculate it manually, but if number of train batches is reported, why not report number of eval batches?

Second, validation happens  after epoch_callback is called. Is there a good reason for this? Can this behavior be changed?  Or is there a good way to detect and of the validation? 

Strange in this context is that in the console the validation data is reported after every epoch:
INFO:root:Epoch[1] Validation-accuracy=0.982595
INFO:root:Epoch[1] Validation-cross-entropy=0.107147

Can it be [made] simply accessible via the API? (along with same values for training set)

Thank you so much!
Alex

**Comment by jrosebr1** [comment]
+1 on this. I would love to have an easy method to grab **both** the training and validation loss **and** accuracy after each epoch so I can plot it.

**Comment by CNevd** [comment]
u can use mx.metric.CompositeEvalMetric() to get multi-metric

**Comment by alinagithub** [comment]
Hello !
Same issue. I tried a custom callback providing the model ref and iterators, calculating the accuracies inside. Unfortunately this does not work if providing the same iterator as used by the fit process - what I understand - but not working even if I provide a separate (copy) of the iterator - for an unknown reason, this freeze the fit process...

Have you found a workaround since you posted this issue? 
Many thanks,

**Comment by undertherain** [comment]
No, my "work-around" so far is to use other frameworks :(

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
